kie-remote/kie-remote-client/src/main/java/org/kie/remote/client/ws/KieRemoteWsAuthenticator.java

```diff
@@ -11,54 +11,99 @@
+ * This is a thread-local {@link Authenticator} instance that allows users to make multiple (authenticated) calls to
+ * web services without running into race conditions involving different threads with different authentication information
+ * (user/password).
  */
 public class KieRemoteWsAuthenticator extends Authenticator {
 
     protected static final Logger logger = LoggerFactory.getLogger(KieRemoteWsAuthenticator.class);
-  
-    // Is a static ThreadLocal GC safe?
+
     private static ThreadLocal<PasswordAuthentication> threadLocalPasswordAuthentication = new ThreadLocal<PasswordAuthentication>();
-    private static AtomicBoolean authenticatorSet = new AtomicBoolean(false);
-   
-    public void setUserAndPassword(String userName, String password) { 
-       if( ! authenticatorSet.get() ) { 
-           // here: 2+ threads, Authenticator not yet set
-           synchronized(authenticatorSet) { 
-               // only 1 thread because of sync block
-               if( authenticatorSet.compareAndSet(false, true) ) { 
-                   Authenticator.setDefault(this);
-               }
-           }
-       }
-       PasswordAuthentication pwdAuth = threadLocalPasswordAuthentication.get();
-       if( pwdAuth != null ) { 
-           logger.debug("Replacing password authentication for user '{}' with new authentication for user '{}'", pwdAuth.getUserName(), userName);
-       } 
-       pwdAuth = new PasswordAuthentication(userName, password.toCharArray());
-       threadLocalPasswordAuthentication.set(pwdAuth);
+    private static KieRemoteWsAuthenticator _instance = new KieRemoteWsAuthenticator();
+
+
+    public static KieRemoteWsAuthenticator getInstance() {
+        return _instance;
     }
-    
-    public void clearUserAndPassword() { 
+
+    private KieRemoteWsAuthenticator() {
+        Authenticator.setDefault(this);
+    }
+
+    public void setUserAndPassword( String userName, String password ) {
+        disableHttpUrlConnectionAuthCache();
+
+        Authenticator.setDefault(_instance);
+        PasswordAuthentication pwdAuth = threadLocalPasswordAuthentication.get();
+        if( pwdAuth != null ) {
+            logger.debug("Replacing password authentication for user '{}' with new authentication for user '{}'",
+                    pwdAuth.getUserName(), userName);
+        }
+        pwdAuth = new PasswordAuthentication(userName, password.toCharArray());
+        threadLocalPasswordAuthentication.set(pwdAuth);
+    }
+
+    public void clearUserAndPassword() {
         threadLocalPasswordAuthentication.set(null);
     }
-    
+
     @Override
-    protected PasswordAuthentication getPasswordAuthentication() { 
+    protected PasswordAuthentication getPasswordAuthentication() {
         return threadLocalPasswordAuthentication.get();
     }
-    
+
+    private static AtomicBoolean authenticatorSet = new AtomicBoolean(false);
```

Comment:
might be worth renaming the variable as it's not used for authenticator being set anymore, though a minor thing